src/HasByNonDependentSubqueryMacro.php

```diff
@@ -146,10 +147,40 @@ protected function adjustArgumentTypeOfOptionalConstraints(callable $constraint,
     {
         $reflection = ReflectionCallable::from($constraint);
 
-        return $reflection->getNumberOfParameters() > 0
+        $type = $reflection->getNumberOfParameters() > 0
             && ($parameter = $reflection->getParameters()[0])->hasType()
-            && \is_a($parameter->getType()->getName(), Builder::class, true)
-            ? $relation->getQuery()
-            : $relation;
+            ? $parameter->getType()
+            : null;
+
+        if (!$type) {
+            return $relation;
+        }
+
+        return $this->determineArgumentType($type, $relation);
+    }
+
+    /**
+     * @param  \ReflectionType                                                                        $type
+     * @param  \Illuminate\Database\Eloquent\Relations\Relation                                       $relation
+     * @return \Illuminate\Database\Eloquent\Builder|\Illuminate\Database\Eloquent\Relations\Relation
+     */
+    protected function determineArgumentType(ReflectionType $type, Relation $relation)
+    {
+        // https://www.php.net/manual/en/class.reflectionnamedtype.php
+        if (method_exists($type, 'getName')) {
+            return \is_a($type->getName(), Builder::class, true)
+                ? $relation->getQuery()
+                : $relation;
+        }
+
+        // https://www.php.net/manual/en/class.reflectionuniontype.php
+        if (method_exists($type, 'getTypes')) {
+            /** @var ReflectionType $firstType */
+            $firstType = $type->getTypes()[0];
+
+            return $this->determineArgumentType($firstType, $relation);
+        }
```

Comment:
Could these `method_exists()` assertions be replaced into `instanceof` ones?

Reply:
I'd love to, but if you use this code in PHP `<= 8.0` it won't know what a `ReflectionUnionType` is. I could do an instance check for the `ReflectionNamedType`, but this is more consisten.

Reply:
I could try to use `is_a($type, 'ReflectionUnionType ', true)` and see if PHP 7 doesn't have a problem with that (trying to resolve the string).

Reply:
In PHP, the right operand of `instanceof` won't be actually required its existence. I'm sure it's okay in any versions.

Reply:
The tests seem to run with PHP 7.4, but PHPStorm rightly shows a warning.

![image](https://user-images.githubusercontent.com/12373573/132952926-823e8197-21ca-4a10-bc46-9175966f122c.png)

Reply:
Would you insert a suppression comment for the warnig only for the single line?

Reply:
How would I do that?